### DISCUSSION TITLE
Can MSW replace a development proxy (maybe in combination with another tool)?

### DISCUSSION THREAD

**Discussion by joeattardi** [mswjs/msw#1501] [Q&A] [answered]
My current project has a development proxy server that runs Webpack middleware. We run it locally and give it a URL of a real backend, which the API requests will be forwarded to. We can define mocks in JSON files; when a request hits the proxy it checks for a matching mock, and if it finds one returns that. Otherwise it proxies the request up to the real backend.

I'm wondering if this whole flow can be replaced with MSW. The server code is a little (unnecessarily) complex, mocks are inflexible as they're just static JSON files. I saw another post where it was mentioned that proxying is not really the intended use of MSW but it's somewhat possible.

I'm wondering if anyone knows of, or uses, another service worker based tool for proxying requests which can be used in tandem with MSW for the mocking piece.

Would _love_ to be able to ditch this server for a great in-browser solution.

**Comment by kettanaito** [comment, marked as answer]
Hey, @joeattardi. 

From what you've described, you can totally use MSW to achieve such a proxy implementation. Any request handler you define has the ability to passthrough the request, which means it will be performed as-is.

```js
rest.get('/user', (req, res, ctx) => {
  if (foundInMocks(req)) {
    return res(ctx.json(fromMocks))
  }

  return req.passthrough()
})
```

> [`req.passthrough()`](https://mswjs.io/docs/api/request/passthrough)

I also agree with your point that keeping mocks as static JSON fixtures isn't the best approach. Describe your server-side behavior in MSW's request handlers and use `req.passthrough()` anytime to perform the request as-is (send it to the actual backend). And thus you've got yourself a development proxy for both browser and Node.js. 

Let me know if you have any questions.

**Reply by joeattardi** [reply]
Thanks this is great! This gets me most of the way there. What I'm not sure about is if I can also use MSW to proxy requests to a _remote_ backend that's different than the host serving my UI.

For example, I run a Webpack dev server on port 3000, which is proxying requests to a real server running on a remote system. So a request to `https://localhost:3000/api/whatever` really goes to `https://myotherserver/api/whatever`.

Is this something MSW can do as well? This is the part I wasn't sure about...

**Reply by kettanaito** [reply]
Sure. You decide what response to return from the request handler. Since `req.passthrough` will perform the request as-is, you need to construct a proxy request to the remote server instead:

```js
rest.get('/resource', (req, res, ctx) => {
  if (condition) {
    const url = new URL(req.url)
    // Construct a correct proxy request URL.
    const proxyUrl = new URL(`${url.path}${url.search}`, 'https://myotherserver.api')
    // Fetch the proxy request and return its response from the handler.
    return ctx.fetch(proxyUrl)
  }
})
```

**Reply by joeattardi** [reply]
This sounds awesome and I will definitely try this out. Thank you!
